### DISCUSSION TITLE
Como pegar o Numero de Serie dos dispositivos usando flutter

### DISCUSSION THREAD

**Discussion by vjmvagnao** [Flutterando/forum#76] [Perguntas e Respostas]
Sabendo que houve alterações nas restriçoes do Android a partir da versão 10, como fazer para pegar o Serial Number do dispositivo, pois tenho necessidade de um indentificador unico para o dispositivo.

https://developer.android.com/about/versions/10/privacy/changes

 Verificando a documentãção é mencionado algo sobre READ_PRIVILEGED_PHONE_STATE, mas como fazer isso com flutter?

**Comment by eduardoflorence** [comment]
@vjmvagnao pela documentação está escrito que:
"esta é uma permissão privilegiada que só pode ser concedida a aplicativos pré-carregados no dispositivo"
Então você não conseguirá fazer isto via Flutter.
Ao invés de pegar o serial number você não poderia trabalhar com o ID do dispositivo?
Se sim, use esse package: https://pub.dev/packages/device_info_plus

**Comment by vjmvagnao** [comment]
Boa tarde @eduardoflorence agradecido, mas fiz isso via device_info_plus, o problema é que o ID muda conforme a aplicação ".apk" é reinstalada no tablet e isso não pode ocorrer. Tipo se eu gero um apk de um PC XXX o devide ID é um se gero de outro PC YYY o Device ID é outro, no mesmo dispositivo.

**Reply by eduardoflorence** [reply]
Isso também aconteceria, caso o APP estivesse na loja e o usuário desinstalasse e instalasse de novo. O que você tem que fazer é pedir as credenciais do usuário toda vez que esse ID fosse alterado para permitir o uso do APP. Nesse momento você iria conferir se ele ainda assina o seu serviço e atualizaria o ID no seu backend. Caso o usuário tente utilizar algum outro dispositivo com o ID antigo, você também teria condições de barrar o uso.
Atualmente nem Android e nem iOS recomendam que você recupere um serial único do aparelho.

**Reply by vjmvagnao** [reply]
@eduardoflorence obrigado pelos esclarecimentos, mas ainda sim preciso pegar o serial number que nunca deve ser alterado, pois os dispositivos são controlados pelo numero de série.

**Comment by G-campos** [comment]
@vjmvagnao e se utilizar o endereço mac do aparelho android ?

**Reply by eduardoflorence** [reply]
Dispositivos com Android 10 (API de nível 29) e versões mais recentes relatam endereços MAC aleatórios para todos os apps que não são proprietários do dispositivo.

**Comment by vjmvagnao** [comment]
@G-campos acredito que cai nas mesmas retrições a partir da versão 10 do android. 

Restrição de identificadores de dispositivo não reconfiguráveis
A partir do Android 10, os apps precisam ter a permissão privilegiada READ_PRIVILEGED_PHONE_STATE para acessar os identificadores não reconfiguráveis do dispositivo, que incluem o IMEI e o número de série.

**Comment by FredericoWP** [comment]
@vjmvagnao , acredito que a galera já disse tudo!
Conforme a documentação que menciona este tema diz.:

`O Android 10 (API de nível 29) adiciona restrições para identificadores não redefiníveis, que incluem IMEI e número de série. Seu app precisa ser um [dispositivo ou um app do proprietário do perfil](https://source.android.com/devices/tech/admin/managed-profiles#device_administration), ter [permissões especiais para operadoras](https://source.android.com/devices/tech/config/uicc) ou ter a permissão privilegiada READ_PRIVILEGED_PHONE_STATE para acessar esses identificadores.`

Maiores detalhes pode ser vistos no link: [Práticas recomendadas para identificadores exclusivos](https://developer.android.com/training/articles/user-data-ids)
